### DISCUSSION TITLE
Feature Request: Modify resize by aspect to include size as well

### DISCUSSION THREAD

**Discussion by fmw42** [ImageMagick/ImageMagick#7193] [Development]
### ImageMagick version

7.1.1.29

### Operating system, version and so on

OSX Ventura

### Description

I would like to have the aspect ratio resize include the desired output size. For example, say we have an image that is 400x300 and I want to resize to aspect 1:1 but size 200x200

`magick image.suffix -resize 200:200! output.suffix`

This would be equivalent to

`magick image.suffix -resize 1:1 -resize 200x200! output.suffix`

So rather than taking the aspect ratio and computing the new size, the ! symbol means compute the aspect ratio from the WxH, then resize by computed aspect but in combination with the desired size.  So basically we compute the dimensions from the aspect ratio as before, but scale those by the desired size. 
This avoids two resamples of the image.

Note without the !, it should work as before.

Example:

Input (400x299)

![barn](https://github.com/ImageMagick/ImageMagick/assets/13458218/64ad264a-5827-4c01-9129-17f7216decf7)

So rather than

`magick barn.jpg -resize 1:1 x.jpg`

or

`magick barn.jpg -resize 200:200 x.jpg`

which both currently give

![x](https://github.com/ImageMagick/ImageMagick/assets/13458218/6ced9d8f-b81a-446f-81f7-0d76f6825278)

which is 299x299

I would like the equivalent of

`magick barn.jpg -resize 1:1 -resize 200x200! y.jpg`

![y](https://github.com/ImageMagick/ImageMagick/assets/13458218/55b1d7ec-42c0-48d0-a812-497539f1b125)

which is 200x200

**Comment by snibgo** [comment]
I can't see how `-resize 200:200!` would be different to `-resize 200x200!`. Wouldn't they give identical results?

**Reply by fmw42** [reply]
-resize 200x200! would distort to reshape from any other aspect ratio.  I suggest something that does what -resize 1:1 -resize 200x200! would do.  First it resize to the best size to fit 1:1 and then resizes to 200x200, but in one resize.

**Comment by urban-warrior** [comment]
Your suggestion steps on a number of design principles:

1. Don't overload an operator, in this case geometries (1:1, 200x200!)
2. Don't have more than one way to do one thing
3. Keep operations atomic

If you get the desired results with `-resize 1:1 -resize 200x200!`, we recommend you use that to maintain the design principles we listed above.

**Reply by fmw42** [reply]
It does not step on 200x200 because it is 200:200!  The colon is the difference. But any other symbol apart from ! is OK, too.

It avoided two resamples. Which is the main issue!

It is more atomic to have one resize than two.

**Comment by snibgo** [comment]
Sorry, I still can't understand the difference.

>  First it resize to the best size to fit 1:1 and then resizes to 200x200, but in one resize.

If we resize to exactly 200 by 200 with `-resize 200x200!`, then any resize before that makes no difference to the final result (apart from rounding errors). The result image will be the input image resized to 200x200 pixels.

Can you mock-up an example of the expected result of `-resize 200:200!`, showing how that is different from the result of `-resize 200x200!`?

**Reply by fmw42** [reply]
Sure.  -resize 200:200 is the same currently as -resize 1:1

`magick barn.jpg -resize 1:1 xx.jpg`

![xx](https://github.com/ImageMagick/ImageMagick/assets/13458218/850eba6b-21e0-422a-9dcf-dd8df3097c27)

`magick barn.jpg -resize 200:200 yy.jpg`

![yy](https://github.com/ImageMagick/ImageMagick/assets/13458218/d5ed8d87-d99b-4fb6-858d-e11e291c3da3)

**Comment by fmw42** [comment]
As the developer has some objections, I will close this and develop the math for %[fx:...] use in -resize to make it one -resize.  I would have had to do that anyway as proof of concept before it could be developed directly into Imagemagick.

**Comment by fmw42** [comment]
Here is the algorithm to do this (Unix syntax)

Suppose we want to take the barn image (aspect of 400:299 = 4:3) and change the aspect to 200:100 = 2:1  resizing to 200x200.  We basically need to crop to the desired aspect, then resize.

![barn](https://github.com/ImageMagick/ImageMagick/assets/13458218/3bc06260-de4a-4b55-8c1e-33a2780d0dd7)


```
aspect="200:100"
ww=`echo "$aspect" | cut -d: -f1`
hh=`echo "$aspect" | cut -d: -f2`
aspect=`im7 magick xc: -format "%[fx:$ww/$hh]" info:`
magick barn.jpg -gravity center -crop "%[fx:($aspect>w/h)?w:h*$aspect]x%[fx:($aspect>w/h)?w/$aspect:h]+0+0" +repage -resize ${ww}x${hh}! x.png

```

![x](https://github.com/ImageMagick/ImageMagick/assets/13458218/95c4fbcf-d304-4242-bcac-5b534083bd0b)

**Comment by snibgo** [comment]
Hmm. From your last post, your proposed algorithm is equivalent to a `-crop` followed by a `-resize`, not one `resize` followed by another `-resize`.

`-crop` responds to the colon `:` symbol, so your crop can be simplified to:
```
-crop ${ww}:${hh}+0+0
```

**Reply by fmw42** [reply]
Thanks. I forgot about the colon in the crop.

**Reply by fmw42** [reply]
OK, thanks to snibgo, the aspect ratio crop followed by the resize, does the job without two resizes.

```
aspect="200:100"
ww=`echo "$aspect" | cut -d: -f1`
hh=`echo "$aspect" | cut -d: -f2`
magick barn.jpg -gravity center -crop ${ww}:${hh}+0+0 +repage -resize ${ww}x${hh}! x.png

```

![x](https://github.com/ImageMagick/ImageMagick/assets/13458218/f8d414ce-59a2-4137-a76d-a7791ce4655c)
